overlord/snapshotstate/snapshotstate_test.go

```diff
@@ -0,0 +1,662 @@
+	shotfile, err := os.Create(filepath.Join(c.MkDir(), "foo.zip"))
+	c.Assert(err, check.IsNil)
+	defer shotfile.Close()
+	fakeIter := func(_ context.Context, f func(*backend.Reader) error) error {
+		c.Assert(f(&backend.Reader{
+			// not wanted
+			Snapshot: client.Snapshot{SetID: setID, Snap: "a-snap"},
+			File:     shotfile,
+		}), check.IsNil)
+
+		return nil
+	}
+	defer snapshotstate.MockBackendIter(fakeIter)()
+
+	snaps, files, err := snapshotstate.SnapNamesInSnapshotSet(setID, []string{"b-snap"})
+	c.Assert(err, check.Equals, client.ErrSnapshotSnapsNotFound)
+	c.Check(snaps, check.IsNil)
+	c.Check(files, check.IsNil)
+}
+
+func (snapshotSuite) TestCheckConflict(c *check.C) {
+	st := state.New(nil)
+	st.Lock()
+	defer st.Unlock()
+	chg := st.NewChange("some-change", "...")
+	tsk := st.NewTask("some-task", "...")
+	tsk.SetStatus(state.DoingStatus)
+	chg.AddTask(tsk)
+
+	// no snapshot state
+	err := snapshotstate.CheckSnapshotChangeConflict(st, 42, "some-change")
+	c.Assert(err, check.ErrorMatches, "internal error: task \"some-task\" of change 1 is missing snapshot information")
+
+	// wrong snapshot state
+	tsk.Set("snapshot", "hello")
+	err = snapshotstate.CheckSnapshotChangeConflict(st, 42, "some-change")
+	c.Assert(err, check.ErrorMatches, "internal error.*")
+
+	tsk.Set("snapshot", map[string]int{"set-id": 42})
+
+	err = snapshotstate.CheckSnapshotChangeConflict(st, 42, "some-change")
+	c.Assert(err, check.ErrorMatches, "snapshot set #42 has a .some-change. change in progress")
+
+	// no change with that label
+	c.Assert(snapshotstate.CheckSnapshotChangeConflict(st, 42, "some-other-change"), check.IsNil)
+
+	// no change with that snapshot id
+	c.Assert(snapshotstate.CheckSnapshotChangeConflict(st, 43, "some-change"), check.IsNil)
+
+	// no non-ready change
+	tsk.SetStatus(state.DoneStatus)
+	c.Assert(snapshotstate.CheckSnapshotChangeConflict(st, 42, "some-change"), check.IsNil)
+}
+
+func (snapshotSuite) TestSaveChecksSnapnamesError(c *check.C) {
+	defer snapshotstate.MockSnapstateAll(func(*state.State) (map[string]*snapstate.SnapState, error) {
+		return nil, errors.New("bzzt")
+	})()
+
+	st := state.New(nil)
+	st.Lock()
+	defer st.Unlock()
+	_, _, _, err := snapshotstate.Save(st, nil, nil)
+	c.Check(err, check.ErrorMatches, "bzzt")
+}
+
+func (snapshotSuite) TestSaveChecksSnapstateConflictError(c *check.C) {
+	defer snapshotstate.MockSnapstateCheckChangeConflictMany(func(*state.State, []string, func(*state.Task) bool) error {
+		return errors.New("bzzt")
+	})()
+
+	st := state.New(nil)
+	st.Lock()
+	defer st.Unlock()
+	_, _, _, err := snapshotstate.Save(st, nil, nil)
+	c.Check(err, check.ErrorMatches, "bzzt")
+}
+
+func (snapshotSuite) TestSaveChecksSetIDError(c *check.C) {
+	st := state.New(nil)
+	st.Lock()
+	defer st.Unlock()
+
+	st.Set("snapshot-set", 42)
+
+	_, _, _, err := snapshotstate.Save(st, nil, nil)
+	c.Check(err, check.ErrorMatches, ".* could not unmarshal .*")
+}
+
+func (snapshotSuite) TestSaveNoSnaps(c *check.C) {
```

Comment:
Maybe `TestSaveNoSnasInState` to differentiate between the case when Save() is called with an empty state and save is called with no-snaps  in Save(..., snapNames,...) ?